### ISSUE TITLE
Interrupt condition for FAISS efficient filtering logic

### ISSUE THREAD

**Issue by neo-anderson** [opensearch-project/k-NN#1898]
I've been reviewing the efficient filtering logic of FAISS HNSW as described in the OpenSearch documentation: https://opensearch.org/docs/latest/search-plugins/knn/filter-search-knn/

## Observations and Questions

1. **Redundant Condition Check**
   The logic includes a condition `R < K and P >= K`. However, if I understand correctly, `P` should always be greater than `K` when the flow reaches this point. This makes the `P >= K` check redundant.

2. **Potential Performance Impact**
   If the `R < K` check is performed after conducting an ANN search on the entire graph, it could lead to frequent "full" graph searches. This is because, as per the flow chart, the condition is evaluated only after the search is complete.

## Proposed Improvement
- Remove `P >= K` condition to avoid confusion
- Check the number of matching candidates DURING the search. If the count reaches or exceeds `K`, interrupt the search.
- Also implement a search interrupt mechanism like the one mentioned in the Lucene's filter implementation to avoid expensive full graph searches

This could potentially reduce unnecessary computations and improve overall efficiency.

**Comment by neo-anderson** [comment]
Not sure if this qualifies as a bug. GitHub is not letting me remove the label. Sorry

**Comment by navneet1v** [comment]
> The logic includes a condition R < K and P >= K. However, if I understand correctly, P should always be greater than K when the flow reaches this point. This makes the P >= K check redundant.

Yes the check is redundant. But its a integer comparison, so shouldn't harm the latency. But I do agree it can be removed. @neo-anderson do you want to fix this condition.


> Not sure if this qualifies as a bug. GitHub is not letting me remove the label. Sorry

Yes its not a bug. But I have removed the bug label.


> Potential Performance Impact
If the R < K check is performed after conducting an ANN search on the entire graph, it could lead to frequent "full" graph searches. This is because, as per the flow chart, the condition is evaluated only after the search is complete.

Yes I agree this is a performance improvement. I did consider this during the implementation but Faiss doesn't have this capability. So we need to build it in FAISS. @neo-anderson do you have any thoughts on how it can be added in FAISS.
